joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/datastructures/ProgramSummary.scala

```diff
@@ -0,0 +1,253 @@
+    */
+  protected val aliasedTypes = mutable.HashMap.empty[String, String]
+
+  /** Given a type name or alias, attempts to resolve its full name using the types currently in scope.
+    *
+    * @param typeName
+    *   the shorthand name.
+    * @return
+    *   the type meta-data if found.
+    */
+  def tryResolveTypeReference(typeName: String): Option[T] = {
+    typesInScope
+      .collectFirst {
+        case typ if typ.name.endsWith(typeName)                                           => typ
+        case typ if aliasedTypes.contains(typeName) && typ.name == aliasedTypes(typeName) => typ
+      }
+      .flatMap(typ => summary.namespaceFor(typ).map(namespace => typ))
+  }
+
+  /** Given the type full name and call name, will attempt to find the matching entry.
+    *
+    * @param typeFullName
+    *   the base type full name.
+    * @param callName
+    *   the call name.
+    * @param argTypes
+    *   the observed argument types. Only relevant for languages that implement overloading.
+    * @return
+    *   the method full name if found.
+    */
+  def tryResolveMethodInvocation(typeFullName: String, callName: String, argTypes: List[String]): Option[M] = {
+    tryResolveTypeReference(typeFullName).flatMap { t =>
+      t.methods.find { m => m.name == callName && isOverloadedBy(m, argTypes) }
+    }
+  }
+
+  /** Determines if, by observing the given argument types, that the method's signature is a plausible match to the
+    * observed arguments.
+    *
+    * The default implementation only considers that the same number of arguments are added and does not account for
+    * variadic arguments nor polymorphism.
+    *
+    * @param method
+    *   the method meta data.
+    * @param argTypes
+    *   the observed arguments from the call-site.
+    * @return
+    *   true if the method could be overloaded by a call with these argument types.
+    */
+  protected def isOverloadedBy(method: M, argTypes: List[String]): Boolean = {
+    method.parameterTypes.size == argTypes.size
+  }
+
+  /** Appends known types imported into the scope.
+    * @param namespace
+    *   the fully qualified imported namespace.
+    */
+  def addImportedNamespace(namespace: String): Unit = {
+    val knownTypesFromNamespace = summary.typesUnderNamespace(namespace)
+    typesInScope.addAll(knownTypesFromNamespace)
+  }
+
+  /** Appends known types imported into the scope.
+    * @param typeOrModule
+    *   the type name or full name.
+    */
+  def addImportedTypeOrModule(typeOrModule: String): Unit = {
+    val matchingTypes = summary.matchingTypes(typeOrModule)
+    typesInScope.addAll(matchingTypes)
+  }
+
+}
+
+/** An implementation of combining the typed scoping structures to manage the available type information at namespace
+  * levels.
+  *
+  * @param summary
+  *   the program summary.
+  */
+class DefaultTypedScope[I, V, S <: TypedScopeElement, M <: MethodLike, F <: FieldLike, T <: TypeLike[M, F]](
```

Comment:
This is very heavy in type parameters - can you please give them more meaningful names? Like with regular parameters, single-char names are ok in very concise functions, i.e. if the scope is very small. That's not the case here, so I'd be helpful to the reader if they have proper names. 

To be honest, I don't want people to see this and blame it all on "Scala is sooo complicated", throw it all away and rewrite it in go. I've seen it happen before, trust me :)

Reply:
Good advice, I'll see what I can do to reduce this and make it a lot friendlier, even with maybe some default scope elements

Reply:
Another option that could help simplify things in some situations might be to use type members instead of type parameters:
```
class A1[B]
// vs
class A2 { type B }
```

Reply:
I've pushed some new changes to address reducing the type arguments, including modelling module-like scope behaviour. 

I still figure `DefaultTypedScope` should be seen as more of a reference implementation, instead of a canonical way to use these tools, but I've chosen to design it in a way that now it should suite most usages I've observed.